File: jar/src/main/java/jline/api/moment/Moment_binomial_from_factorial.java
package jline.api.moment;

import jline.util.matrix.Matrix;

/**
 * Binomial moments from factorial moments.
 *
 * <p>Converts the factorial moments f_n = E[N(N-1)...(N-n+1)] of a discrete random variable N into the binomial moments b_n = E[binom(N,n)] via the one-to-one correspondence
 *
 * <pre>
 *   b_n = f_n / n!
 * </pre>
 *
 * <p>Reference:
 * A. Heindl and A. van de Liefvoort. Moment conversions for discrete
 * distributions. PMCCS, 2003, Section 4.
 *
 * @since LINE 3.0
 */
public final class Moment_binomial_from_factorial {
    private Moment_binomial_from_factorial() {}

    /**
     * Binomial moments from factorial moments.
     *
     * @param f column vector of length n+1 holding the moments of order 0..n,
     *          element 0 being the order-0 moment, which equals 1
     * @return column vector of length n+1 holding the converted moments
     */
    public static Matrix moment_binomial_from_factorial(Matrix f) {
        int n = f.length() - 1;
        Matrix out = new Matrix(n + 1, 1);
        double fact = 1.0;
        for (int i = 0; i <= n; i++) {
            if (i > 0) {
                fact = fact * i;
            }
            out.set(i, 0, f.get(i) / fact);
        }
        return out;
    }
}

File: matlab/src/api/moment/moment_binomial_from_factorial.m
function b = moment_binomial_from_factorial(f)
% b = moment_binomial_from_factorial(f)
%
% Converts the factorial moments f_n = E[N(N-1)...(N-n+1)] of a discrete
% random variable N into the binomial moments b_n = E[nchoosek(N,n)] via the
% one-to-one correspondence
%
%   b_n = f_n / n!
%
% Input:
%   f: vector of length n+1 holding f_0,...,f_n, i.e. f(i) is the moment of
%      order i-1 and f(1) = f_0 = 1
%
% Output:
%   b: vector of length n+1 holding b_0,...,b_n, with the same orientation
%      as f
%
% Reference:
% A. Heindl and A. van de Liefvoort. Moment conversions for discrete
% distributions. PMCCS, 2003, Section 4.
%
% Example:
% b = moment_binomial_from_factorial([1,2,4,8])

fcol = f(:);
n = length(fcol)-1;
b = zeros(n+1,1);
for i = 0:n
    b(i+1) = fcol(i+1) / factorial(i);
end
if isrow(f)
    b = b.';
end
end
